src/Keboola/S3Extractor/S3AsyncDownloader.php

```diff
@@ -0,0 +1,117 @@
+use Aws\ResultInterface;
+use Aws\Exception\AwsException;
+use Psr\Log\LoggerInterface;
+use Retry\RetryProxy;
+use Retry\Policy\SimpleRetryPolicy;
+use Retry\BackOff\ExponentialBackOffPolicy;
+use GuzzleHttp\Psr7\LazyOpenStream;
+use function Keboola\Utils\formatBytes;
+
+class S3AsyncDownloader
+{
+    private const MAX_ATTEMPTS = 5;
+    private const INTERVAL_MS = 500;
+    private const MAX_CONCURRENT_DOWNLOADS = 50;
+
+    /**
+     * @var S3Client
+     */
+    private $client;
+
+    /**
+     * @var LoggerInterface
+     */
+    private $logger;
+
+    /**
+     * @var CommandInterface[]
+     */
+    private $commands = [];
+
+    /**
+     * @var array
+     */
+    private $keys = [];
+
+    /**
+     * @var int
+     */
+    private $downloadedSize = 0;
+
+    /**
+     * @param S3Client $client
+     * @param LoggerInterface $logger
+     */
+    public function __construct(S3Client $client, LoggerInterface $logger)
+    {
+        $this->client = $client;
+        $this->logger = $logger;
+    }
+
+    /**
+     * @param array $fileParameters
+     * @return void
+     */
+    public function addFileRequest(array $fileParameters): void
+    {
+        $this->commands[] = $this->client->getCommand('getObject', $fileParameters);
+    }
+
+    /**
+     * @return void
+     */
+    public function processRequests(): void
+    {
+        (new RetryProxy(
+            new SimpleRetryPolicy(self::MAX_ATTEMPTS),
+            new ExponentialBackOffPolicy(self::INTERVAL_MS),
+            $this->logger
+        ))->call(function () {
+            $this->makeCommandPool()
+                ->promise()
+                ->then(function () {
+                    $this->logger->info(sprintf(
+                        'Downloaded %d file(s) (%s)',
+                        count($this->commands),
+                        formatBytes($this->downloadedSize)
+                    ));
+                })
+                ->wait();
+        });
+    }
+
+    /**
+     * @return CommandPool
+     */
+    private function makeCommandPool(): CommandPool
+    {
+        return new CommandPool($this->client, $this->commands, [
+            'concurrency' => self::MAX_CONCURRENT_DOWNLOADS,
+            'before' => function (CommandInterface $command, int $index) {
+                $this->keys[$index] = $command->offsetGet('Key');
+            },
+            'fulfilled' => function (ResultInterface $result, int $index) {
+                $body = $result->get('Body');
+                /** @var LazyOpenStream $body */
+                $fileSize = $body->getSize();
+                $this->logger->info(sprintf(
+                    'Downloaded file complete /%s (%s)',
+                    $this->keys[$index],
+                    formatBytes($body->getSize())
+                ));
+
+                $this->downloadedSize += $fileSize;
+            },
+            'rejected' => static function (AwsException $reason) {
+                throw $reason;
```

Comment:
Pokud tohle chápu dobře, tak když jeden z těch 30 tisíc souborů vyfailuje, tak se dělá retry všech, že? Nešlo by si ty soubory, co vyfailují, hodit bokem a stáhnout je znova? Z nich zase může samozřejmě nějakej vyfailovat :-)

U uploadu do S3 jsem to dělal tak, že na začátku je pole promisů, u kterých se ve smyčce čeká na jejich splnění. Jak se resolvují, tak ty, které vyfailují, se znovu zařadí do stejného pole promisů a ta stejný cyklus resolve/wait jede znova (u uploadu to ještě umí předat state, takže i z jednoho souboru to uploaduje pouze tu část, co se ještě neodeslala). 

https://github.com/keboola/storage-api-php-client/blob/40a17b968b6c4378457a7c01641b5dfab2e72827/src/Keboola/StorageApi/S3Uploader.php#L140-L167

Reply:
Jo, včera mi to večer taky cvaklo, upraveno a doloženo testem `RetryDownloadFileTest::testRetrySuccess()`